File: lab/adversary/canary_scan.py
#!/usr/bin/env python3
"""Scan files and packet captures for planted canary strings.

Usage: canary_scan.py CANARIES_FILE OUT_JSONL PATH [PATH ...]
CANARIES_FILE: one canary per line, "id<TAB>text".
A PATH may be a file or a directory (walked). Files ending .pcap are also parsed:
TCP streams are reassembled per direction and UDP payloads extracted, and those
are scanned in addition to the raw file bytes.

Each canary is searched in these forms: UTF-8, UTF-16LE, UTF-16BE, hex (both
cases), URL-encoded, JSON \\u-escaped, and base64 at all three alignments. Any
gzip or zlib stream found in a blob is decompressed and the result scanned too,
and base64 runs are decoded and rescanned (two levels), as are JSON string literals
containing \\u escapes. Every hit is written to
OUT_JSONL and flushed. Exit status 1 if anything was found, 0 if clean.
"""
import base64, binascii, json, os, re, struct, sys, urllib.parse, zlib

def variants(text):
    b = text.encode("utf-8")
    v = {"utf8": b, "utf16le": text.encode("utf-16-le"), "utf16be": text.encode("utf-16-be"),
         "hex": binascii.hexlify(b), "HEX": binascii.hexlify(b).upper(),
         "url": urllib.parse.quote(text).encode(), "json_u": "".join(f"\\u{ord(c):04x}" for c in text).encode()}
    for pad in range(3):   # base64 of the canary at each alignment; keep the stable middle
        enc = base64.b64encode(b"\0" * pad + b)
        skip = (pad * 4 + 2) // 3 + (1 if pad else 0)
        core = enc[skip: len(enc) - 4]
        if len(core) >= 8: v[f"b64_{pad}"] = core
    return v

def decoded_children(blob):
    """Yield (how, bytes) for compressed streams and base64 runs inside blob."""
    for m in re.finditer(rb"\x1f\x8b\x08", blob):
        try:
            d = zlib.decompressobj(16 + zlib.MAX_WBITS); yield ("gzip", d.decompress(blob[m.start():]))
        except zlib.error: pass
    for m in re.finditer(rb"\x78[\x01\x5e\x9c\xda]", blob):
        try:
            d = zlib.decompressobj(); out = d.decompress(blob[m.start():])
            if out: yield ("zlib", out)
        except zlib.error: pass
    # JSON string literals containing \u escapes: encoders escape only some
    # characters, so decode the literal instead of matching one escaped form.
    for m in re.finditer(rb'"(?:[^"\\]|\\.){0,4096}?\\u[0-9a-fA-F]{4}(?:[^"\\]|\\.){0,4096}"', blob):
        try: yield ("json", json.loads(m.group(0).decode("latin-1")).encode("utf-8", "surrogatepass"))
        except (ValueError, UnicodeError): pass
    for m in re.finditer(rb"[A-Za-z0-9+/]{16,}={0,2}", blob):
        s = m.group(0)
        s = s[: len(s) - len(s) % 4] if not s.endswith(b"=") else s
        try: yield ("base64", base64.b64decode(s, validate=True))
        except (binascii.Error, ValueError): pass

def scan_blob(blob, canaries, where, emit, depth=0):
    for cid, forms in canaries.items():
        for form, needle in forms.items():
            i = blob.find(needle)
            if i >= 0:
                emit({"where": where, "canary": cid, "form": form, "offset": i})
    # Two levels: base64 wrapping gzip is a common way plaintext hides.
    if depth < 2:
        for how, child in decoded_children(blob):
            scan_blob(child, canaries, f"{where}|{how}", emit, depth + 1)

def pcap_streams(data):
    """Classic libpcap: return {flow: bytes} for TCP (reassembled) and a list of UDP payloads."""
    if len(data) < 24: return {}, []
    magic = data[:4]
    endian = "<" if magic in (b"\xd4\xc3\xb2\xa1", b"\x4d\x3c\xb2\xa1") else ">"
    linktype = struct.unpack(endian + "I", data[20:24])[0]
    off, segs, udp = 24, {}, []
    while off + 16 <= len(data):
        _, _, incl, _ = struct.unpack(endian + "IIII", data[off:off+16]); off += 16
        pkt = data[off:off+incl]; off += incl
        if linktype == 1:  eth, l3 = pkt[12:14], pkt[14:]            # Ethernet
        elif linktype == 113: eth, l3 = pkt[14:16], pkt[16:]         # Linux cooked
        elif linktype == 276: eth, l3 = pkt[0:2], pkt[20:]           # Linux cooked v2
        else: continue
        if eth == b"\x08\x00" and len(l3) >= 20:
            ihl = (l3[0] & 15) * 4; proto = l3[9]; src, dst = l3[12:16], l3[16:20]; l4 = l3[ihl:]
        elif eth == b"\x86\xdd" and len(l3) >= 40:
            proto = l3[6]; src, dst = l3[8:24], l3[24:40]; l4 = l3[40:]
        else: continue
        if proto == 6 and len(l4) >= 20:
            sp, dp, seq = struct.unpack("!HHI", l4[:8]); doff = (l4[12] >> 4) * 4; payload = l4[doff:]
            if payload: segs.setdefault((src, sp, dst, dp), {})[seq] = payload
        elif proto == 17 and len(l4) >= 8:
            udp.append(l4[8:])
    streams = {}
    for flow, d in segs.items():
        streams[flow] = b"".join(d[s] for s in sorted(d))   # retransmissions share a seq and collapse
    return streams, udp

def main():
    cfile, outp, paths = sys.argv[1], sys.argv[2], sys.argv[3:]
    canaries = {}
    for line in open(cfile, encoding="utf-8"):
        if line.strip():
            cid, text = line.rstrip("\n").split("\t", 1); canaries[cid] = variants(text)
    out = open(outp, "w"); hits = []
    def emit(h):
        hits.append(h); out.write(json.dumps(h) + "\n"); out.flush()
    files = []
    for p in paths:
        if os.path.isdir(p):
            for root, _, names in os.walk(p):
                files += [os.path.join(root, n) for n in names]
        elif os.path.exists(p): files.append(p)
    for f in sorted(files):
        data = open(f, "rb").read()
        scan_blob(data, canaries, f, emit)
        if f.endswith(".pcap"):
            streams, udp = pcap_streams(data)
            for flow, s in streams.items():
                scan_blob(s, canaries, f"{f}#tcp", emit)
            for i, u in enumerate(udp):
                scan_blob(u, canaries, f"{f}#udp{i}", emit)
    out.close()
    print(json.dumps({"files": len(files), "hits": len(hits)}), flush=True)
    sys.exit(1 if hits else 0)

if __name__ == "__main__":
    main()

File: lab/tool/lib/canary_scan.dart
/// Scan files and packet captures for planted canary strings.
///
/// Each canary is searched in these forms: UTF-8, UTF-16LE, UTF-16BE, hex
/// (both cases), URL-encoded, JSON \u-escaped, and base64 at all three
/// alignments. Gzip and zlib streams found in a blob are decompressed, base64
/// runs are decoded, and JSON string literals containing \u escapes are decoded;
/// each result is scanned too, to two levels (base64 wrapping gzip is a common
/// way plaintext hides). Files ending .pcap are also parsed: TCP streams are
/// reassembled per direction and UDP payloads extracted.
library;

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

/// One place a canary was found.
class Hit {
  Hit(this.where, this.canary, this.form, this.offset);
  final String where;
  final String canary;
  final String form;
  final int offset;
  Map<String, Object> toJson() =>
      {'where': where, 'canary': canary, 'form': form, 'offset': offset};
}

/// Python's urllib.parse.quote with its default safe="/".
String _quote(String s) {
  final out = StringBuffer();
  for (final b in utf8.encode(s)) {
    final c = String.fromCharCode(b);
    if (RegExp(r'[A-Za-z0-9_.\-~/]').hasMatch(c)) {
      out.write(c);
    } else {
      out.write('%${b.toRadixString(16).toUpperCase().padLeft(2, '0')}');
    }
  }
  return out.toString();
}

List<int> _utf16(String s, {required bool little}) {
  final out = <int>[];
  for (final u in s.codeUnits) {
    final hi = u >> 8, lo = u & 0xff;
    out.addAll(little ? [lo, hi] : [hi, lo]);
  }
  return out;
}

String _hex(List<int> b) =>
    b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();

/// Every encoded form searched for one canary.
Map<String, List<int>> variants(String text) {
  final b = utf8.encode(text);
  final v = <String, List<int>>{
    'utf8': b,
    'utf16le': _utf16(text, little: true),
    'utf16be': _utf16(text, little: false),
    'hex': ascii.encode(_hex(b)),
    'HEX': ascii.encode(_hex(b).toUpperCase()),
    'url': ascii.encode(_quote(text)),
    'json_u': ascii.encode(text.codeUnits
        .map((c) => '\\u${c.toRadixString(16).padLeft(4, '0')}')
        .join()),
  };
  for (var pad = 0; pad < 3; pad++) {
    // base64 of the canary at each alignment; keep the stable middle.
    final enc = ascii.encode(base64.encode([...List.filled(pad, 0), ...b]));
    final skip = (pad * 4 + 2) ~/ 3 + (pad > 0 ? 1 : 0);
    if (enc.length - 4 - skip >= 8) {
      v['b64_$pad'] = enc.sublist(skip, enc.length - 4);
    }
  }
  return v;
}

/// Index of [needle] in [hay], or -1.
int indexOfBytes(List<int> hay, List<int> needle) {
  if (needle.isEmpty || needle.length > hay.length) return -1;
  final first = needle[0];
  outer:
  for (var i = 0; i <= hay.length - needle.length; i++) {
    if (hay[i] != first) continue;
    for (var j = 1; j < needle.length; j++) {
      if (hay[i + j] != needle[j]) continue outer;
    }
    return i;
  }
  return -1;
}

/// Inflate from [start], tolerating trailing data after the stream ends, the
/// way Python's decompressobj does. Returns what was decoded, or null.
List<int>? _inflate(List<int> blob, int start, {required bool gzip}) {
  final filter = RawZLibFilter.inflateFilter(gzip: gzip);
  final out = <int>[];
  try {
    const chunk = 512;
    for (var i = start; i < blob.length; i += chunk) {
      final end = i + chunk < blob.length ? i + chunk : blob.length;
      filter.process(blob, i, end);
      List<int>? o;
      while ((o = filter.processed(flush: false)) != null) {
        out.addAll(o!);
      }
    }
    List<int>? o;
    while ((o = filter.processed()) != null) {
      out.addAll(o!);
    }
  } on Exception {
    // Trailing bytes after the compressed stream, or not a stream at all.
  }
  return out.isEmpty ? null : out;
}

final _b64Run = RegExp('[A-Za-z0-9+/]{16,}={0,2}');
final _jsonLit = RegExp(
    r'"(?:[^"\\]|\\.){0,4096}?\\u[0-9a-fA-F]{4}(?:[^"\\]|\\.){0,4096}"');

/// Yield (how, bytes) for compressed streams, JSON literals and base64 runs.
Iterable<(String, List<int>)> decodedChildren(List<int> blob) sync* {
  for (var i = 0; i + 2 < blob.length; i++) {
    if (blob[i] == 0x1f && blob[i + 1] == 0x8b && blob[i + 2] == 0x08) {
      final d = _inflate(blob, i, gzip: true);
      if (d != null) yield ('gzip', d);
    }
    if (blob[i] == 0x78 && const [0x01, 0x5e, 0x9c, 0xda].contains(blob[i + 1])) {
      final d = _inflate(blob, i, gzip: false);
      if (d != null) yield ('zlib', d);
    }
  }
  final text = latin1.decode(blob);
  for (final m in _jsonLit.allMatches(text)) {
    try {
      final s = jsonDecode(m.group(0)!) as String;
      yield ('json', utf8.encode(s));
    } on FormatException {
      // not a JSON literal after all
    }
  }
  for (final m in _b64Run.allMatches(text)) {
    var s = m.group(0)!;
    if (!s.endsWith('=')) s = s.substring(0, s.length - s.length % 4);
    try {
      yield ('base64', base64.decode(s));
    } on FormatException {
      // not base64
    }
  }
}

/// Scan one blob, then what can be decoded out of it, to two levels.
void scanBlob(List<int> blob, Map<String, Map<String, List<int>>> canaries,
    String where, void Function(Hit) emit,
    [int depth = 0]) {
  canaries.forEach((cid, forms) {
    forms.forEach((form, needle) {
      final i = indexOfBytes(blob, needle);
      if (i >= 0) emit(Hit(where, cid, form, i));
    });
  });
  if (depth < 2) {
    for (final (how, child) in decodedChildren(blob)) {
      scanBlob(child, canaries, '$where|$how', emit, depth + 1);
    }
  }
}

/// Classic libpcap: TCP streams reassembled per direction, and UDP payloads.
(Map<String, List<int>>, List<List<int>>) pcapStreams(Uint8List data) {
  final segs = <String, Map<int, List<int>>>{};
  final udp = <List<int>>[];
  if (data.length < 24) return ({}, udp);
  final bd = ByteData.sublistView(data);
  final m = data.sublist(0, 4);
  final little = (m[0] == 0xd4 && m[1] == 0xc3) || (m[0] == 0x4d && m[1] == 0x3c);
  final e = little ? Endian.little : Endian.big;
  final linktype = bd.getUint32(20, e);
  var off = 24;
  while (off + 16 <= data.length) {
    final incl = bd.getUint32(off + 8, e);
    off += 16;
    if (off + incl > data.length) break;
    final pkt = data.sublist(off, off + incl);
    off += incl;
    List<int> eth;
    Uint8List l3;
    if (linktype == 1 && pkt.length > 14) {
      eth = pkt.sublist(12, 14); l3 = pkt.sublist(14);
    } else if (linktype == 113 && pkt.length > 16) {
      eth = pkt.sublist(14, 16); l3 = pkt.sublist(16);
    } else if (linktype == 276 && pkt.length > 20) {
      eth = pkt.sublist(0, 2); l3 = pkt.sublist(20);
    } else {
      continue;
    }
    int proto;
    List<int> src, dst;
    Uint8List l4;
    if (eth[0] == 0x08 && eth[1] == 0x00 && l3.length >= 20) {
      final ihl = (l3[0] & 15) * 4;
      proto = l3[9]; src = l3.sublist(12, 16); dst = l3.sublist(16, 20);
      if (ihl > l3.length) continue;
      l4 = l3.sublist(ihl);
    } else if (eth[0] == 0x86 && eth[1] == 0xdd && l3.length >= 40) {
      proto = l3[6]; src = l3.sublist(8, 24); dst = l3.sublist(24, 40);
      l4 = l3.sublist(40);
    } else {
      continue;
    }
    final l4d = ByteData.sublistView(l4);
    if (proto == 6 && l4.length >= 20) {
      final sp = l4d.getUint16(0), dp = l4d.getUint16(2), seq = l4d.getUint32(4);
      final doff = (l4[12] >> 4) * 4;
      if (doff > l4.length) continue;
      final payload = l4.sublist(doff);
      if (payload.isNotEmpty) {
        final flow = '${_hex(src)}:$sp>${_hex(dst)}:$dp';
        // Retransmissions share a sequence number and collapse.
        (segs[flow] ??= {})[seq] = payload;
      }
    } else if (proto == 17 && l4.length >= 8) {
      udp.add(l4.sublist(8));
    }
  }
  final streams = <String, List<int>>{};
  segs.forEach((flow, d) {
    final keys = d.keys.toList()..sort();
    streams[flow] = [for (final k in keys) ...d[k]!];
  });
  return (streams, udp);
}

/// Scan [paths] (files or directories) for the canaries in [canaryFile]
/// ("id<TAB>text" per line). Every hit is appended to [out] and flushed.
Future<List<Hit>> scanPaths(
    String canaryFile, List<String> paths, IOSink out) async {
  final canaries = <String, Map<String, List<int>>>{};
  for (final line in File(canaryFile).readAsLinesSync()) {
    if (line.trim().isEmpty) continue;
    final tab = line.indexOf('\t');
    canaries[line.substring(0, tab)] = variants(line.substring(tab + 1));
  }
  final files = <String>[];
  for (final p in paths) {
    if (FileSystemEntity.isDirectorySync(p)) {
      files.addAll(Directory(p)
          .listSync(recursive: true)
          .whereType<File>()
          .map((f) => f.path));
    } else if (File(p).existsSync()) {
      files.add(p);
    }
  }
  files.sort();
  final hits = <Hit>[];
  void emit(Hit h) {
    hits.add(h);
    out.writeln(jsonEncode(h.toJson()));
  }

  for (final f in files) {
    final data = File(f).readAsBytesSync();
    scanBlob(data, canaries, f, emit);
    if (f.endsWith('.pcap')) {
      final (streams, udp) = pcapStreams(data);
      streams.forEach((flow, s) => scanBlob(s, canaries, '$f#tcp', emit));
      for (var i = 0; i < udp.length; i++) {
        scanBlob(udp[i], canaries, '$f#udp$i', emit);
      }
    }
    await out.flush();
  }
  return hits;
}
